onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -15,23 +15,48 @@ namespace onnxruntime {
    etc.
    */
 
-// ComputeOffset: compute offset into a tensor. This is essentially the dot-product of
-// index and stride, restricted to the specified number of axes.
-static inline size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t>& stride, int64_t num_axes) {
-  size_t offset = 0;
-  for (int64_t j = 0; j < num_axes; ++j) {
-    offset += index[j] * stride[j];
+typedef struct MultiIndex {
+  size_t index;
+  size_t upper_bound;
+  int64_t stride;
+  MultiIndex() {}
+  void Init(size_t i, size_t n, int64_t s) {
+    index = i;
+    upper_bound = n;
+    stride = s;
   }
-  return offset;
+} MultiIndex;
+
+static size_t IncrementIndexAndComputeOffsetSetup(MultiIndex* mindex, int64_t num_axes, const std::vector<int64_t>& target_dims,
+                                                  const std::vector<size_t>& stride, size_t element_size) {
+  size_t naxes = 0;
+  for (int64_t i = 0; i < num_axes; ++i) {
+    if (target_dims[i] == 1)
+      continue;
+    mindex[naxes].Init(0, static_cast<size_t>(target_dims[i]), stride[i] * element_size);
+    ++naxes;
+  }
+  ORT_ENFORCE(naxes > 0, "Method IncrementIndexAndComputeOffset assumes this value is strictly positive.");
+  return naxes;
 }
 
-// IncrementIndex: Increment an index into a tensor (in lexicographic ordering), wrapping
-// around the specified upper_bound.
-static inline void IncrementIndex(std::vector<int64_t>& index, const std::vector<int64_t>& upper_bound, int64_t num_axes) {
-  for (int64_t k = num_axes - 1; k >= 0; --k) {
-    index[k]++;
-    if (index[k] < upper_bound[k]) break;
-    index[k] = 0;
+// Combines multi-index increment and corresponding pointer in the tensor to transpose.
+template <typename T>
+static void IncrementIndexAndComputeOffset(MultiIndex* mindex, size_t naxes, const T*& local_source) {
+  MultiIndex* it = (mindex + naxes) - 1;
+  local_source += it->stride;
+  if (++it->index < it->upper_bound)
```

Comment:
It would really help to have some comments on a number of lines here explaining the logic as there's a lot of intrinsic knowledge required to understand the steps. Would help detect potential issues as well.

Reply:
I added more comments. I'll make some benchmark today to measure the impact of the instructions ORT_ENFORCE inside the loops involving indices.